### PR TITLE
sysfs: workaround for bogus NUMA meminfo entry.

### PR DESCRIPTION
**Pull request by klihub**
On some HW we've seen more free that total memory being reported.
This causes an unsigned integer underflow in the calculation of
the amount of memory being used, resulting in exorbitant amounts
being reported. As a workaround, we calculate the amount of used
memory ourselves, capping the free memory to the total available
at max. to avoid similar underflows.

### CODE REVIEW

File: pkg/sysfs/system.go [hunk 1]
```diff
@@ -891,7 +891,6 @@ func (n *node) MemoryInfo() (*MemInfo, error) {
 		map[string]interface{}{
 			"MemTotal:": &buf.MemTotal,
 			"MemFree:":  &buf.MemFree,
-			"MemUsed:":  &buf.MemUsed,
```
Comment (jukkar) [inline pkg/sysfs/system.go:894]: Why do we remove this? It is not obvious from the commit.

Comment (klihub) [reply pkg/sysfs/system.go:894]: Because I fsck'd up the commit...

Comment (klihub) [reply pkg/sysfs/system.go:894]: Updated with a hopefully less b0rken fix.

**Review by jukkar — APPROVED** [review]

**Review by askervin — DISMISSED** [review]
LGTM

File: pkg/sysfs/system.go [hunk 2]
```diff
@@ -910,6 +909,20 @@ func (n *node) MemoryInfo() (*MemInfo, error) {
...
+	//
+	// On some HW we've seen more free that total memory being reported.
+	// This causes an unsigned integer underflow in the calculation of
+	// the amount of memory being used, resulting in exorbitant amounts
+	// being reported. As a workaround, we calculate the amount of used
+	// memory ourselves, capping the free memory to the total available
+	// at max. to avoid similar underflows.
+	//
+	if buf.MemFree > buf.MemTotal {
+		buf.MemFree = buf.MemTotal
```
Comment (kad) [inline pkg/sysfs/system.go:922]: I'd say `MemFree = MemTotal - MemUsed` in this case.

**Comment by kad** [comment]
The more I'm thinking about that, the more I'm inclined to report that as an "buggy kernel detected. Please upgrade/fix your kernel".

**Comment by klihub** [comment]
> The more I'm thinking about that, the more I'm inclined to report that as an "buggy kernel detected. Please upgrade/fix your kernel".

Should we report it as an unrecoverable error either with `log.Fatal` (doing an `os.Exit()`) or with `log.Panic()` (`panic()`ing) ?

**Review by jukkar — DISMISSED** [review]
LGTM

**Comment by kad** [comment]
I think we agreed that this is a kernel bug, so workarround might be not good to implement. Shall we close this PR?

**Comment by klihub** [comment]
> I think we agreed that this is a kernel bug, so workarround might be not good to implement. Shall we close this PR?

But a `log.Fatal()` or a `panic()` would still probably be the right course of action.

**Comment by klihub** [comment]
@askervin @jukkar @kad @marquiz PTAL, I updated this to bail out with an error if we run into such a buggy kernel.

**Review by marquiz — COMMENTED** [review]
One typo found.

Completely refusing to start in this scenario feels like a very strong action  Just wondering what if every software piece in the system refused to start if it found out that there's an unfixed kernel bug 😜 But obviously you've thought about it more thoroughly so can go with this

**Comment by klihub** [comment]
> One typo found.
> 
> Completely refusing to start in this scenario feels like a very strong action Just wondering what if every software piece in the system refused to start if it found out that there's an unfixed kernel bug stuck_out_tongue_winking_eye But obviously you've thought about it more thoroughly so can go with this

Well, the original patch tested for the same condition, then capped the free memory to the total available if the former was larger than the latter. I changed it to this bail out (encouraged by the fact that we've never seen this problem again once our downstream customer updated their kernel. At least with the topology-aware policy enabled, even if we didn't bail out here there would be a failure to start later when the policy tries to initialize itself. However, I think it is better to fail contextually as close to detecting the real problem as possible as it usually provides better error context and helps the end user in diagnosing the problem instead of sending him down a wild goose chase.

File: pkg/sysfs/system.go [hunk 3]
```diff
@@ -910,6 +909,23 @@ func (n *node) MemoryInfo() (*MemInfo, error) {
 	if err != nil {
 		return nil, err
 	}
+
+	//
+	// On some HW we've seen more free that total memory being reported.
```
[the remark this answers is not in the document]
Comment (klihub) [reply pkg/sysfs/system.go]: And there was that `whcih` that went unnoticed by either of us... I rewrote the comment and the commit message to be cleared... hopefully without typos this time.

**Review by marquiz — APPROVED** [review]
lgtm

**Pull request merged**